rpc/storage.go

```diff
@@ -0,0 +1,273 @@
+	}
+	result.ContractsStorageProofs, err = getContractsStorageProofs(stateReader, storageKeys)
+	if err != nil {
+		return nil, ErrInternal.CloneWithData(err)
+	}
+
+	return result, nil
+}
+
+// StorageKeys represents an item in `contracts_storage_keys. parameter
+// https://github.com/starkware-libs/starknet-specs/blob/647caa00c0223e1daab1b2f3acc4e613ba2138aa/api/starknet_api_openrpc.json#L938
+type StorageKeys struct {
+	Contract felt.Felt   `json:"contract_address"`
+	Keys     []felt.Felt `json:"storage_keys"`
+}
+
+// MerkleNode represents a proof node in a trie
+// https://github.com/starkware-libs/starknet-specs/blob/647caa00c0223e1daab1b2f3acc4e613ba2138aa/api/starknet_api_openrpc.json#L3632
+type MerkleNode interface{}
+
+// https://github.com/starkware-libs/starknet-specs/blob/647caa00c0223e1daab1b2f3acc4e613ba2138aa/api/starknet_api_openrpc.json#L3644
+type MerkleBinaryNode struct {
+	Left  *felt.Felt `json:"left"`
+	Right *felt.Felt `json:"right"`
+}
+
+// TODO[pnowosie]: link to specs
+type MerkleEdgeNode struct {
+	Path   *felt.Felt `json:"path"`
+	Length int        `json:"length"`
+	Child  *felt.Felt `json:"child"`
+}
+
+// TODO[pnowosie]: link to specs
+type MerkleLeafNode struct {
+	Value *felt.Felt `json:"value"`
+}
+
+// HashToNode represents an item in `NODE_HASH_TO_NODE_MAPPING` specified here
+// https://github.com/starkware-libs/starknet-specs/blob/647caa00c0223e1daab1b2f3acc4e613ba2138aa/api/starknet_api_openrpc.json#L3667
+type HashToNode struct {
+	Hash *felt.Felt `json:"node_hash"`
+	Node MerkleNode `json:"node"`
+}
+
+// TODO[pnowosie]: link to specs
+type LeafData struct {
+	Nonce     *felt.Felt `json:"nonce"`
+	ClassHash *felt.Felt `json:"class_hash"`
+}
+
+// TODO[pnowosie]: link to specs
+type ContractProof struct {
+	Nodes      []*HashToNode `json:"nodes"`
+	LeavesData []*LeafData   `json:"contract_leaves_data"`
+}
+
+// TODO[pnowosie]: link to specs
+type GlobalRoots struct {
+	ContractsTreeRoot *felt.Felt `json:"contracts_tree_root"`
+	ClassesTreeRoot   *felt.Felt `json:"classes_tree_root"`
+	BlockHash         *felt.Felt `json:"block_hash"`
+}
+
+// TODO[pnowosie]: link to specs
+type StorageProofResult struct {
+	ClassesProof           []*HashToNode   `json:"classes_proof"`
+	ContractsProof         *ContractProof  `json:"contracts_proof"`
+	ContractsStorageProofs [][]*HashToNode `json:"contracts_storage_proofs"`
+	GlobalRoots            *GlobalRoots    `json:"global_roots"`
+}
+
+func getClassesProof(reader core.StateReader, classes []felt.Felt) ([]*HashToNode, error) {
+	ctrie, _, err := reader.ClassTrie()
+	if err != nil {
+		return nil, err
+	}
+	result := []*HashToNode{}
+	for _, class := range classes {
+		nodes, err := getProof(ctrie, &class)
+		if err != nil {
+			return nil, err
+		}
+		result = append(result, nodes...)
+	}
+	return result, nil
+}
+
+func getContractsProof(reader core.StateReader, contracts []felt.Felt) (*ContractProof, error) {
+	strie, _, err := reader.StorageTrie()
```

Comment:
why don't we handle the closer functions?

Reply:
This is read-only and closer tries to update the root if it had changed.
BTW for the StateReader (r-o) I should drop the closer fn from the response but later for the snap-sync we will use it to put ranges into tries...

Reply:
Good point, maybe we can remove it since it's not being used, and to prevent any potential problems in the future?

Reply:
I will leave for now and try to refactor once me merge the snap-sync pr (the same methods with same return types)